python/mxnet/ndarray/ndarray.py

```diff
@@ -155,6 +155,13 @@ def _new_alloc_handle(shape, ctx, delay_alloc, dtype=mx_real_t):
             ctypes.c_int(int(_DTYPE_NP_TO_MX[np.dtype(dtype).type])),
             ctypes.byref(hdl)))
     else:
+        size = 1
+        for idx in shape:
+            size = size * idx
+        if size > 2**31:
+            raise Exception("[_new_alloc_handle] Size of tensor you are trying to allocate is " +
```

Comment:
Shall we not just let the backend handle that logic and bubble up the exception? The frontend shouldn't care about these kinds of things I'd say

Reply:
Feel free to let me know if you'd like me to elaborate

Reply:
actually if you look at the C API call below it is passing shape as `c_array_buf(mx_uint, native_array('I', shape)),` so the code i a lot of cases break here it self saying that integer value passed overflows unit32 range. Therefore, a check here is unavoidable. Let me know if this makes sense. If you have a suggestion to make this check better, I am happy to hear it :)

Reply:
Makes sense, thanks for elaborating. Would you mind adding a comment?

Reply:
As proposed by Frank in the design review, it would be better to check int64 related condition inside C API, right? I agree with Marco, we should minimize the backend logic checking in frontend.

Reply:
@apeforest the call will fail here itself. It won't even go to backend as stated in my comment above: https://github.com/apache/incubator-mxnet/pull/16570/files#r338846832

Reply:
> Makes sense, thanks for elaborating. Would you mind adding a comment?

Sure !

Reply:
Didn't we have 32 and 64 bit C api functions?

It sounds counter intuitive initially, but if we have overloaded c api functions we could always expose them. If we're compiled without 64bit Support, the c api would then return an error instead of passing on the request. IMO this design makes the typing a bit more explicit and transparent (while I have to admit that I generally dislike the verbose explicity in c++) instead of hiding all the magic in a simple typedef override and some preprocessor statements.

The logic in the frontend would then be to call the 32bit c api functions with small tensors and 64bit one with large tensors. Of course it needs some tweaking because that's not a fail proof design against overflow of the intermediary results.

What it seems like to me is that a C Api is supposed to be a static contract that is always present. But with the 64bit Support we're now introducing versions of the c api that depend on the compile options. This then breaks the exchangability and decoupling. In the end it's a matter of taste, but I think we should uphold ourselves to a standard where the c api never changes but the backend might behave differently depending on the chosen options. At the same time, we might want to make it a rule that c api function definitions must not be conditional. 

It's a bit derailing, but what do you think about that idea?

Reply:
`The logic in the frontend would then be to call the 32bit c api functions with small tensors and 64bit one with large tensors. Of course it needs some tweaking because that's not a fail proof design against overflow of the intermediary results.`

A: Actually its not that straight, the operator code still needs to use 32 bit and 64 bit versions. Since the performance of both versions of operators in not the same using 64bit version of both versions of C API is not the ideal choice.

`But with the 64bit Support we're now introducing versions of the c api that depend on the compile options. This then breaks the exchangability and decoupling. In the end it's a matter of taste, but I think we should uphold ourselves to a standard where the c api never changes but the backend might behave differently depending on the chosen options`

A: This is possible. We can have both 32 bit and 64 bit version of all the operators in the the code. This would require duplicate versions of same code which involves a lot of inline and templatized code. This can be used to dynamically switch between 32bit and 64bit versions of the operators keeping all the C APIs as 64bit. This was considered earlier but as you yourself already know have pointed out in many PRs it would lead to blowing up of the generated binaries `libmxnet.so` so I decided not to move ahead with this approach.

Hope this addresses your concerns